sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+
+import contextlib
+import functools
+import time
+import uuid
+from datetime import datetime
+
+import pytz
+from future.moves import queue
+
+import apache_beam as beam
+import apache_beam.io.gcp.pubsub as beam_pubsub
+from apache_beam import coders
+from apache_beam.io.gcp.pubsub import PubsubMessage
+from apache_beam.io.gcp.pubsub import ReadFromPubSub
+from apache_beam.io.gcp.pubsub import WriteToPubSub
+from apache_beam.runners.direct.direct_runner import _DirectWriteToPubSubFn
+from apache_beam.runners.interactive.caching import pcollectioncache
+from apache_beam.testing import datatype_inference
+from apache_beam.transforms import PTransform
+from apache_beam.transforms.window import TimestampedValue
+from apache_beam.utils.timestamp import Timestamp
+
+try:
+  from weakref import finalize
+except ImportError:
+  from backports.weakref import finalize
+
+try:
+  from google import api_core
+  from google.cloud import pubsub
+  from google.api_core import exceptions as gexc
+except ImportError:
+  api_core = None
+  pubsub = None
+  gexc = None
+
+__all__ = [
+    "StreamBasedCache",
+    "PubSubBasedCache",
+]
+
+
+class StreamBasedCache(pcollectioncache.PCollectionCache):
+  pass
+
+
+class PubSubBasedCache(StreamBasedCache):
+
+  _reader_class = ReadFromPubSub
+  _writer_class = WriteToPubSub
+  _reader_passthrough_arguments = {
+      "id_label",
+      "with_attributes",
+      "timestamp_attribute",
+  }
+  requires_coder = True
+
+  _default_timestamp_attribute = "ts"
+
+  def __init__(self, location, mode="error", persist=False, **writer_kwargs):
+    self.location = location
+    self.coder = writer_kwargs.get("coder")
+    self._writer_kwargs = writer_kwargs
+    self._writer_kwargs["timestamp_attribute"] = self._writer_kwargs.get(
+        "timestamp_attribute", self._default_timestamp_attribute)
+    self._persist = persist
+    self._timestamp = 0
+    self._finalizers = []
+    self._child_subscriptions = []
+    self._primary_sub_exhausted = False
+
+    if mode not in ["error", "append", "overwrite"]:
+      raise ValueError("mode must be set to 'error' or 'append'.")
+
+    # Initialize PubSub resources. Note that we cannot save pub_client and
+    # sub_client as class attributes because they fail to serialize using
+    # pickle.
+    pub_client = pubsub.PublisherClient()
```

Comment:
Ideally the code that reads and writes to/from Pub/Sub would be reused. Specifically I'm thinking about the DirectRunner code, like the `rfc3339_or_milli` code.

Reply:
I agree. However, I think it may be better to leave this as a TODO to be addressed in a future refactoring? Since InteractiveRunner is alpha, the code for caching to PubSub is likely to undergo several iterations. If those changes also affect DirectRunner, it may increase the surface area for introducing bugs or at least slow the development cycle?